yoyodyne/models/hmm.py

```diff
@@ -0,0 +1,418 @@
+"""Hard monotonic neural HMM classes."""
+
+import argparse
+from typing import Callable, Dict, Optional, Union
+
+import numpy
+import torch
+from torch import nn
+
+from .. import data, defaults
+from . import lstm, modules
+
+
+class HardAttentionHmm(lstm.LSTMEncoderDecoder):
+    """Hard Attention Transducer.
+
+    Learns probability distribution of target string by modeling transduction
+    of source string to target string as Markov process. Assumes each character
+    produced is conditioned by state transitions over each source character.
+
+    Default model assumes independence between state and non-monotonic
+    progression over source string. `enforce_monotonic` enforces monotonic
+    state transition (model progresses over each source character), and
+    `attention_context` allows conditioning of state transition over previous
+    _n_ states. 
+    
+    After:
+        Wu, S. and Cotterell, R. 2019. Exact hard monotonic attention for
+        character-level transduction. In _Proceedings of the 57th Annual
+        Meeting of the Association for Computational Linguistics_, pages
+        1530-1537.
+
+    Original implementation:
+        https://github.com/shijie-wu/neural-transducer
+    """
+
+    enforce_monotonic: bool
```

Comment:
While these are optional in the sense that you don't have to pass them in, they're not nullable, so they shouldn't be labeled `Optional`. I took the liberty of fixing this.